### PR TITLE
feat(auth, apple): Game Center sign-in support

### PR DESCRIPTION
**Pull request by russellwheatley**
## Description

I've had to change the way we get the AuthCredential on Apple platforms. Game Center credential is different to the other providers. It requires a callback to receive. See link: https://firebase.google.com/docs/auth/ios/game-center#authenticate_with_firebase
Here is the snippet:
```objc
// Get Firebase credentials from the player's Game Center credentials
[FIRGameCenterAuthProvider getCredentialWithCompletion:^(FIRAuthCredential *credential,
                                                         NSError *error) {
  // The credential can be used to sign in, or re-auth, or link or unlink.
  if (error == nil) {
    [[FIRAuth auth] signInWithCredential:credential
                              completion:^(FIRUser *user, NSError *error) {
      // If error is nil, player is signed in.
    }];
  }
}];
```

Previously, we called a synchronous method to get the credential: https://github.com/firebase/flutterfire/blob/master/packages/firebase_auth/firebase_auth/ios/Classes/FLTFirebaseAuthPlugin.m#L641C33-L641C66 

But I had to make this a callback to allow for Game Center credential retrieval: https://github.com/firebase/flutterfire/pull/12228/files#diff-926b26fe3f9119fb3548066f1b05028306ff8766f9a04abba0fefdb5dac5c0b2R642-R645

This method is used for `signInWithCredential`, `linkWithCredential()`, `reauthenticateWithCredential()` & `updatePhoneNumber()` API which will allow the user to call these methods `signInWithCredential`, `linkWithCredential()`, `reauthenticateWithCredential()` with Game Center credential.

## Testing Game Center sign-in

1. Add Game Center capability on Xcode project which also brings in framework for Game Player.
2. Create an app in "App Store Connect" and select the "Game Center" option to allow authentication.
3. Implement the following in AppDelegate to login with Game Center:
```objc
__weak GKLocalPlayer *localPlayer = [GKLocalPlayer localPlayer];
localPlayer.authenticateHandler = ^(UIViewController *gcAuthViewController,
                                    NSError *error) {
  if (gcAuthViewController != nil) {
    // Pause any activities that require user interaction, then present the
    // gcAuthViewController to the player.
  } else if (localPlayer.isAuthenticated) {
    // Player is signed in to Game Center. Can now login with Game Center credential
	NSLog(@"logged in with Game Center");
  } else {
    // Error
  }
};
```
4. Now able to login with Game Center with Firebase.

## Tested signInWithCredential(), linkWithCredential() & reauthenticateWithCredential()
Screenshot 2024-01-31 at 09 26 56

## Related Issues

closes https://github.com/firebase/flutterfire/issues/12191
closes https://github.com/firebase/flutterfire/issues/9441

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`).
This will ensure a smooth and quick review process. Updating the `pubspec.yaml` and changelogs is not required.

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [ ] All existing and new tests are passing.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] The analyzer (`melos run analyze`) does not report any problems on my PR.
- [ ] I read and followed the [Flutter Style Guide].
- [ ] I signed the [CLA].
- [ ] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change.
- [X] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/firebase/flutterfire/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

File: packages/firebase_auth/firebase_auth/ios/Classes/FLTFirebaseAuthPlugin.m [hunk 1]
```diff
@@ -1336,6 +1399,17 @@ - (void)signInWithProviderApp:(nonnull AuthPigeonFirebaseApp *)app
                                                 FlutterError *_Nullable))completion {
   FIRAuth *auth = [self getFIRAuthFromAppNameFromPigeon:app];
 
+  if ([signInProvider.providerId isEqualToString:kSignInMethodGameCenter]) {
+    completion(
+        nil,
+        [FlutterError
+            errorWithCode:@"sign-in-failure"
+                  message:
+                      @"Game Center sign-in requires signing in with 'linkWithCredential()' API."
```
Comment (Lyokone) [inline packages/firebase_auth/firebase_auth/ios/Classes/FLTFirebaseAuthPlugin.m:1408]:
```suggestion
                      @"Game Center sign-in requires signing in with 'signInWithCredential()' API."
```

Comment (Lyokone) [reply packages/firebase_auth/firebase_auth/ios/Classes/FLTFirebaseAuthPlugin.m:1408]: Should we have the same error for linkWithProvider?

File: packages/firebase_auth/firebase_auth/ios/Classes/FLTFirebaseAuthPlugin.m [hunk 2]
```diff
@@ -638,18 +639,37 @@ - (FIRAuth *_Nullable)getFIRAuthFromAppNameFromPigeon:(AuthPigeonFirebaseApp *)p
   return auth;
 }
 
-- (FIRAuthCredential *_Nullable)getFIRAuthCredentialFromArguments:(NSDictionary *)arguments
-                                                              app:(AuthPigeonFirebaseApp *)app {
+- (void)getFIRAuthCredentialFromArguments:(NSDictionary *)arguments
+                                      app:(AuthPigeonFirebaseApp *)app
+                               completion:(void (^)(FIRAuthCredential *credential,
+                                                    NSError *error))completion {
```
Comment (Lyokone) [inline packages/firebase_auth/firebase_auth/ios/Classes/FLTFirebaseAuthPlugin.m:645]: I don't see any Pigeon related changes on the Dart part, is there some files missing?

Comment (russellwheatley) [reply packages/firebase_auth/firebase_auth/ios/Classes/FLTFirebaseAuthPlugin.m:645]: Yes, just double checked, it's the same error response as signing in with provider:
```bash
"[auth/operation-not-allowed] - The federated login request is not supported."
```

I'll update the code 👍

File: packages/firebase_auth/firebase_auth/ios/Classes/FLTFirebaseAuthPlugin.m (hunk 1 above)
Comment (russellwheatley) [reply packages/firebase_auth/firebase_auth/ios/Classes/FLTFirebaseAuthPlugin.m:1408]: Yes, just double checked, it's the same error response as signing in with provider:
```bash
"[auth/operation-not-allowed] - The federated login request is not supported."
```

I'll update the code 👍

File: packages/firebase_auth/firebase_auth/ios/Classes/FLTFirebaseAuthPlugin.m (hunk 2 above)
Comment (russellwheatley) [reply packages/firebase_auth/firebase_auth/ios/Classes/FLTFirebaseAuthPlugin.m:645]: This method is internal to the plugin and isn't part of Pigeon API.

File: packages/firebase_auth/firebase_auth/ios/Classes/FLTFirebaseAuthPlugin.m (hunk 1 above)
Comment (russellwheatley) [reply packages/firebase_auth/firebase_auth/ios/Classes/FLTFirebaseAuthPlugin.m]: Done 👍

File: packages/firebase_auth/firebase_auth/ios/Classes/FLTFirebaseAuthPlugin.m (hunk 2 above)
Comment (Lyokone) [reply packages/firebase_auth/firebase_auth/ios/Classes/FLTFirebaseAuthPlugin.m:645]: Oh nice didn't realise thanks :)

**Review by Lyokone — APPROVED** [review]

**Pull request merged**